[4.0] Event - onUserBeforeDataValidation

BrainforgeUK opened the pull request:
Pull Request for Issue #19584 .

### Summary of Changes

Added event onContentValidateData

### Testing Instructions

Create a field plugin which implements onContentValidateData function like this...
  public function onContentValidateData($form, &$data) {
    // Change the data in some way
  }

### Expected result

The data get's changed.

### Actual result

The data get's changed.

### Documentation Changes Required

Add documentation for onContentValidateData event.
Add documentation for the existing, but undocumented and now deprecated, onUserBeforeDataValidation event.
This event will only get triggered when validate() overridden by a user component model.

BrainforgeUK committed:
Improve how XML errors raise an exception. (#1)

On XML errors raise exception in loadFile() method instead of leaving it to getInstance() function. This makes it easier to detect the location of the error during development.

BrainforgeUK committed:
Added trigger event onContentValidateData (#19584)

BrainforgeUK committed:
Changed curly braces position (#19584)

mbabker commented on libraries/src/CMS/Model/Form.php:
```diff
@@ -332,7 +332,15 @@ public function validate($form, $data, $group = null)
+		$calledClass = get_called_class();
+		if (strpos($calledClass, 'User') === 0)
+		{
+			// onUserBeforeDataValidation retained for backward compatibility
```
Instead of this and trying to only trigger an event based on the calling class, I'd check the dispatcher for listeners for the deprecated event and trigger a deprecation message.

```php

$dispatcher = \JFactory::getContainer()->get('dispatcher');

if (!empty($dispatcher->getListeners('onUserBeforeDataValidation'))) {
    @trigger_error('The `onUserBeforeDataValidation` event is deprecated and will be removed in 5.0, use the `onContentValidateData` event instead.', E_USER_DEPRECATED);

    \JFactory::getApplication()->triggerEvent('onUserBeforeDataValidation', array($form, &$data));
}

\JFactory::getApplication()->triggerEvent('onContentValidateData', array($form, &$data));
```

BrainforgeUK committed:
Changed to check dispatcher for listeners for the deprecated event and if any exist trigger a deprecation message.

See review of Joomla issue  #19584 by https://github.com/mbabker

brianteeman commented:
Please can you look at resolving the conflicts

BrainforgeUK committed:
Resolved pull request conflicts.

BrainforgeUK committed:
Moved closing parenthesis of a multi-line function call to a line by itself

BrainforgeUK committed:
Changed spaces to tabs

BrainforgeUK commented:
Now got a Javascript conflict - any suggestions?

brianteeman commented:
Not sure what you mean

Quy commented:
I think he is referring to the false error with drone.

BrainforgeUK commented:
See lines 556 to 561 of
http://ci.joomla.org/joomla/joomla-cms/7735/5

Also mentioned here, but how can I resolved?
https://github.com/jasmine/jasmine/issues/1486

BrainforgeUK committed:
Improved performance and ordering

Admin control panel very slow on system with a large number of guests. guest column is not indexed, added userid to where clause in order to speed up query by using an index. Also without an order by the order of rows returned is not guaranteed, so adding an order by ensures the latest X rows are returned.

BrainforgeUK committed:
Update joomla.sql (#3)

BrainforgeUK committed:
Changed LEFT JOIN to INNER (#2)

Logged in non-guest will always be users!

BrainforgeUK committed:
Change value column in #__fields_values to longtext (#4)

See issue #19618
[com_fields| Joomla custom field data length

brianteeman commented:
You have added completely unrelated code3 to this issue

You must use a different branch and create a different pull request for each issue you are trying to resolve

In your git client revert this latest change, commit and push

Then create a NEW branch and pull from staging
Make your new changes, commit and push
go to https://github.com/joomla/joomla-cms and you will see info that you can create a new pull request from your new branch

BrainforgeUK committed:
Reverted previous value column in #__fields_value change. (#6)

Pull request error.

sanderpotjer commented:
@BrainforgeUK can you update your PR to solve the conflicts and remove the changes visible on https://github.com/joomla/joomla-cms/pull/20152/files that are not related to the PR? Thanks in advance!

BrainforgeUK commented:
The Delete option is disabled for me in the ... menu for each of those not required.
![delete](https://user-images.githubusercontent.com/3941269/57574714-f42a1c80-7435-11e9-8f23-819d1cc9d9fe.jpg)

The only change required for this issue is the last one listed in form.php
![form php](https://user-images.githubusercontent.com/3941269/57574722-0dcb6400-7436-11e9-88b2-cef358084f81.jpg)

Cannot see how to get round this if the delete menu item is disabled.

Quy commented:
Could it be that the repo is no longer available?

![20152](https://user-images.githubusercontent.com/368084/57649052-1037d600-757c-11e9-9fe3-063b3fc6d1a3.jpg)

roland-d commented:
@Quy Indeed, a new PR should be made as the old branch has been deleted. I am not sure if @BrainforgeUK can restore the original branch. Otherwise it is easier to make a new PR.

BrainforgeUK commented:
Created new pull request.

BrainforgeUK commented:
See pull request #24901
